### PR TITLE
feat: implement context budget layer (#24)

### LINKED ISSUE #24: Implement context management (budgeting, summarization, intent anchor)
## Background
Spec 01 (lines 137-158) describes context management for agent LLM calls:
- Context budgeting: priority-ordered assembly with token limits
- Context summarization: condense old conversation turns
- Intent anchor: prevent agent drift in long-running tasks

Deferred from Phase 6 (knowledge graph) because agents work fine today with simple history loading. Should be designed and specced before implementation.

## Requirements
- Context budget with priority tiers (system prompt > task context > entity memory > bullpen > KG)
- Summarization of conversation history beyond 20 turns
- Intent anchor stored in task record, included in every burst
- Account for variable context windows across providers (Claude 200k vs other models via OpenRouter)

## Dependencies
- ~~Entity memory~~ ✓
- ~~Bullpen~~ ✓
- ~~Per-skill token tracking (#326)~~ ✓ — provides baseline consumption data to inform tier sizing rather than guessing allocations

## Design considerations

Budget tier allocations should be informed by real usage data from #326. Key questions the data should answer before designing budgets:
- How much context do skills typically consume per invocation?
- Which tiers are the most variable in size?
- Where are the biggest opportunities to reclaim space?

If multi-provider support (#379) lands before or alongside this, budget logic must handle different context window sizes per provider.

## Sequencing

- **Blocked by:** #326 (need real usage data to size tiers)
- **Informed by:** #379 (variable context windows across providers)

### PR DESCRIPTION
**Pull request by josephfung**
## Summary

- Adds a token-aware context budgeting layer that wraps the existing priority-ordered assembly in the agent runtime
- Measures per-tier token cost using a local character-ratio estimator (no API calls), enforces inclusion/exclusion based on available budget, and emits `context.budget` bus events for observability
- Tiers that exceed remaining budget are hard-dropped; conversation history supports partial inclusion (truncates oldest turns)
- Budget configuration (`response_reserve`) is per-agent via YAML config; safety margin (5%) is system-wide

## What's included

| Component | Files |
|-----------|-------|
| Token estimator | `src/agents/llm/token-estimator.ts` — character-ratio heuristic, context window map, exhaustiveness guard |
| ContextBudget class | `src/agents/llm/context-budget.ts` — allocateRequired, allocate, allocateHistory (binary search), getReport |
| Bus event | `src/bus/events.ts` — ContextBudgetPayload, ContextBudgetEvent, factory function |
| Agent config | `src/agents/loader.ts`, `agents/coordinator.yaml` — context_budget.response_reserve |
| Runtime integration | `src/agents/runtime.ts` — wraps assembly, emits event, warns on tier drops |
| Bus permissions | `src/bus/permissions.ts` — context.budget event routing |
| Tests | 3 test files, 18+ unit tests + integration test |

## Observability

Every LLM call now emits a `context.budget` event (landing in `audit_log`) with:
- Per-tier token estimates and included/dropped status
- Total budget utilization percentage
- History turn counts (total vs included)
- Warning logs when security-relevant tiers (sender context) or all history is dropped

## Test plan

- [x] Token estimator: string, ContentBlock[], edge cases, context window map
- [x] ContextBudget: construction, allocateRequired, allocate, allocateHistory (full/truncated/dropped), getReport
- [x] Runtime integration: context.budget event emitted with correct payload
- [x] Full test suite passes, tsc --noEmit clean

Closes #24

## Summary by CodeRabbit

* **New Features**
  * Added token-aware context budgeting for agent LLM calls, enabling intelligent management of conversation history within model context limits.
  * Introduced configurable response token reserves per agent to ensure space for model responses.
  * Added context budget observability via new telemetry events tracking budget utilization and per-tier allocation decisions.

* **Documentation**
  * Added design and implementation plans for the context budgeting layer.

* **Tests**
  * Added comprehensive unit tests for token estimation and context budget allocation logic.
  * Added integration tests verifying budget event emission during agent task execution.

[Review Change Stack](https://app.coderabbit.ai/change-stack/josephfung/curia/pull/555)

### CODE REVIEW

**Comment by josephfung** [comment]
@coderabbitai review

File: agents/coordinator.yaml [hunk 1]
```diff
@@ -4,6 +4,8 @@ description: Central coordinator — routes all messages, delegates to specialis
 model:
   provider: anthropic
   model: claude-sonnet-4-6
+context_budget:
+  response_reserve: 8192
```
[the remark this answers is not in the document]
Comment (josephfung) [reply agents/coordinator.yaml:8]: Fixed in fde7538 — added `context_budget` to `schemas/agent-config.schema.json` with `additionalProperties: false` and `response_reserve` as an optional integer. Startup validation now accepts the field.

File: src/agents/loader.ts [hunk 2]
```diff
@@ -55,6 +55,13 @@ export interface AgentYamlConfig {
     max_cost_usd?: number;
     max_errors?: number;
   };
+  /** Context budget config — controls how much of the model's context window is
+   *  reserved for the response. The budgeting layer in runtime.ts uses this to
+   *  enforce tier-based inclusion/exclusion. */
+  context_budget?: {
+    /** Tokens reserved for the model's response. Default: 8192. */
+    response_reserve?: number;
+  };
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/agents/loader.ts:64]: Fixed in the same commit (fde7538) — `context_budget` is now in both the TypeScript interface (`loader.ts`) and the JSON schema (`schemas/agent-config.schema.json`). The two stay in sync.

File: src/agents/runtime.ts [hunk 3]
```diff
@@ -306,15 +315,50 @@ export class AgentRuntime {
...
 
+    const budgetedHistory = ctxBudget.allocateHistory(
+      history.map(t => ({ role: t.role, content: t.content }) as Message),
+    );
+    if (history.length > 0 && budgetedHistory.length === 0) {
+      logger.warn(
+        { agentId, conversationId, historyTurnsAvailable: history.length, remaining: ctxBudget.remaining },
+        'All conversation history dropped by context budget — model will have no conversation context',
+      );
+    }
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/agents/runtime.ts]: Good catch — this was a real priority inversion. Fixed in fde7538.

The allocation order is now:
1. `allocateRequired('system_prompt', ...)` — always included
2. `allocateRequired('user_message', ...)` — reserved upfront so history can't starve it
3. `allocate('sender_context', ...)` — security-relevant, gets priority over history
4. `allocate('bullpen', ...)` — inter-agent coordination
5. `allocateHistory(...)` — takes whatever budget remains, supports partial inclusion

This ensures sender context (with authorization, trust scores, injection risk) is never dropped because history was greedy. History, being the only tier with partial inclusion (truncation), is the natural "flex" tier.

Messages array order is unchanged: system → sender → bullpen → history → user.

File: tests/unit/agents/llm/context-budget.test.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        contextWindow: 300,
+        responseReserve: 50,
+        safetyMargin: 0.0,
+      });
+      const turns = [turn(350), turn(350), turn(350), turn(350)];
+      const included = budget.allocateHistory(turns);
+      expect(included).toHaveLength(2);
+      expect(included[0]).toBe(turns[2]);
+      expect(included[1]).toBe(turns[3]);
+    });
```
[the remark this answers is not in the document]
Comment (josephfung) [reply tests/unit/agents/llm/context-budget.test.ts]: Fair point — renamed both tests in fde7538:

- \"respects minKeep\" → \"greedily includes all turns when budget allows (minKeep is a soft preference)\"
- \"drops all history when even minKeep turns do not fit\" → \"drops all history when no single turn fits the remaining budget\"

\`minKeep\` is preserved in the API signature for forward compatibility but the current implementation is purely greedy. The test names now accurately reflect what they verify.

**Comment by josephfung** [comment]
Addressed the 2 nitpicks from the review as well (both in fde7538):

**Bullpen detection (lines 517-524):** Switched from the fragile `messages[bullpenInsertAt]` index + reference check to `messages.find(m => m.content.startsWith('[Bullpen'))` — matching the same sentinel used by `refreshBullpenContext`. This survives future message reorderings.

**Hard-coded default model (line 319):** Extracted `'claude-sonnet-4-6'` into `DEFAULT_MODEL_NAME` exported from `token-estimator.ts`, living alongside the context window map and `DEFAULT_SAFETY_MARGIN` so the default and the lookup stay in lockstep.

**Comment by josephfung** [comment]
@coderabbitai approve
